### PR TITLE
fix: only reset WebView cookies when the profile changed

### PR DESCRIPTION
**Pull request by criticalAY**
## Purpose / Description
`setDataDirectorySuffix` is unavailable before API 28 so this PR handles that

## Fixes
* Fixes part of #18117

## Approach
See commits

## How Has This Been Tested?
Added unit tests and read documentation for the API

## Learning (optional, can help others)
NA

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/multiprofile/ProfileManager.kt [hunk 1]
```diff
@@ -198,9 +198,26 @@ class ProfileManager private constructor(
...
+     * [WebView.setDataDirectorySuffix] is unavailable before API 28, so profiles
+     * cannot have separate WebView data there. Clearing on every start would
+     * initialize the WebView on the startup path and drop the AnkiWeb session of
+     * users who never switch profiles, so only clear on an actual change.
+     */
+    private fun resetWebViewDataIfProfileChanged(profileId: ProfileId) {
+        if (profileRegistry.getWebViewProfileId() == profileId) return
+
+        Timber.i("WebView data belongs to another profile, clearing cookies")
+        CookieManager.getInstance().removeAllCookies(null)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multiprofile/ProfileManager.kt]: `null` is the callback executed when the cookies are removed.

Define `removeAllCookiesSync { }`, and `setWebViewProfileId` after the call

```kt
     fun CookieManager.removeAllCookiesSync(block: CookieManager.() -> Unit) = this.removeAllCookies { 
        this.flush()
        block()
    }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/multiprofile/ProfileManager.kt [hunk 2]
```diff
@@ -198,9 +198,26 @@ class ProfileManager private constructor(
...
+     * Clears WebView cookies when the active profile differs from the one the
+     * on-disk WebView data belongs to.
+     *
+     * [WebView.setDataDirectorySuffix] is unavailable before API 28, so profiles
+     * cannot have separate WebView data there. Clearing on every start would
+     * initialize the WebView on the startup path and drop the AnkiWeb session of
+     * users who never switch profiles, so only clear on an actual change.
+     */
+    private fun resetWebViewDataIfProfileChanged(profileId: ProfileId) {
+        if (profileRegistry.getWebViewProfileId() == profileId) return
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multiprofile/ProfileManager.kt]: Add a test: I believe - on the first run, `getWebViewProfileId` is not yet set, so `null != profileId` and this would clear data

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt [hunk 3]
```diff
@@ -149,13 +152,65 @@ class ProfileManagerTest {
...
-        val mockCookies = mockk<CookieManager>(relaxed = true)
-        every { CookieManager.getInstance() } returns mockCookies
+    fun `Legacy device clears cookies on first run as no WebView owner is recorded (Pre-API 28)`() {
+        val cookieManager = mockCookieManager()
+        assertNull(prefs.getString(KEY_WEBVIEW_PROFILE_ID, null), "no WebView owner is recorded before the first run")
+
+        ProfileManager.create(context)
+
+        verify(exactly = 1) { cookieManager.removeAllCookies(any()) }
+        assertEquals(ProfileId.DEFAULT.value, prefs.getString(KEY_WEBVIEW_PROFILE_ID, null))
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt:163]: Is this what we want?

Comment (criticalAY) [reply AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt:163]: No, it isn't. That first run is every pre-28 user's first launch once ProfileManager is actually wired in, and their WebView data is already Default's

File: AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt [hunk 4]
```diff
@@ -149,13 +152,85 @@ class ProfileManagerTest {
...
+        ProfileManager.create(context)
+
+        verify(exactly = 0) { cookieManager.removeAllCookies(any()) }
+        verify(exactly = 0) { CookieManager.getInstance() }
+        assertEquals(ProfileId.DEFAULT.value, prefs.getString(KEY_WEBVIEW_PROFILE_ID, null))
+    }
+
+    @Test
+    @Config(sdk = [Build.VERSION_CODES.O_MR1])
+    fun `Legacy device clears cookies once across restarts of the same profile (Pre-API 28)`() {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt]: What does 'restart of the same profile' mean here?

**Comment by david-allison** [comment]
(forgot to comment on the review) 

I /think/ this is good to go and it's just a test name/comment needed

File: AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt (hunk 4 above)
Comment (criticalAY) [reply AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt]: How about
```fun `Legacy device clears cookies only once while the active profile is unchanged (Pre-API 28)`()```

Comment (criticalAY) [reply AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt]: what I meant was the app launches again with `p_bob` still active, the recorded owner already matches, so nothing is cleared the second time.

Comment (david-allison) [reply AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt]: Hmm... maybe `Legacy device does not clear cookies for the same profile repeatedly`

Then a comment/in the code: the first operation clears the cookies, the second operation does not

Comment (criticalAY) [reply AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt]: Renamed and added the comments

File: AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt [hunk 5]
```diff
@@ -149,13 +152,87 @@ class ProfileManagerTest {
...
+        ProfileManager.create(context)
+
+        verify(exactly = 1) { cookieManager.removeAllCookies(any()) }
+        verify(exactly = 1) { cookieManager.flush() }
+        assertEquals("p_bob", prefs.getString(KEY_WEBVIEW_PROFILE_ID, null))
+    }
+
+    @Test
+    @Config(sdk = [Build.VERSION_CODES.O_MR1])
+    fun `Legacy device clears cookies again once the active profile changed (Pre-API 28)`() {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/multiprofile/ProfileManagerTest.kt]: I don't think 'again' is correct here

**Review by david-allison — APPROVED** [review]
I'm nitpicking on test names 🫠, let's get this in

**Comment by criticalAY** [comment]
We can get this in
